### ISSUE TITLE
Question regarding computation of q(z| x, y) in code.

### ISSUE THREAD

**Issue by aayushARM** [mazrk7/gmvae#1]
Hi @mazrk7 ,
Thanks a lot for the code, first of all. I had a question regarding your calculation of q(z|x,y) here: https://github.com/mazrk7/gmvae/blob/master/scripts/gmvae.py#L246
Here, you used the y sampled from qy.
In RuiShi's original code, he uses the ground-truth cluster labels to compute q(z|x, y), as can be seen on this and the next line: https://github.com/RuiShu/vae-clustering/blob/master/gmvae_k.py#L49

Is this your own modification to make the model unsupervised? It doesn't say so in the Readme, so I wanted to confirm it is not a typo or something.

Thanks in advance!
- Ayush.

**Comment by mazrk7** [comment]
Hi @aayushARM,

Thanks for taking interest! 

First of all, I believe both mine and RuiShi's original implementation are _unsupervised_. 

The difference is that in RuiShi's code, he marginalises over all k class values for inference of q(z|x,y), following from [Kingma's M2 model](https://arxiv.org/pdf/1406.5298.pdf) (see Section 3.1.2 in particular). 
In my case, I use the Gumbel-Softmax distribution as a continuous relaxation of the discrete variable to directly allow backpropagation through y ~ q(y|x) (see Section 3.3 of the corresponding [paper](https://arxiv.org/pdf/1611.01144.pdf), which comments on this).

I believe I tested both approaches at some point and found the latter to be better performing. Hope that clarifies things and let me know if you have any other questions :)

Mark

**Comment by aayushARM** [comment]
Thanks for the response, @mazrk7 !
I understand the difference between his and your implementation better now. I also had another quick question regarding this:
I'm interested in using the GMVAE framework for anomaly detection, and for that, the only difference would be that after training,  instead of finding cluster label 'y' for each test image as you do, I need to find the likelihood of the sampled 'z' corresponding to the test image w.r.t all the clusters and then sum them up:

![ll](https://user-images.githubusercontent.com/20727454/95024317-47ff5c00-0650-11eb-9b43-a28ec9fa1097.png)
(here, p(z'|c') is the same as p(z|y) in your code)

The images producing LL lower than a threshold can then be regarded as anomalies. In your code, you have a Fully-connected layer that maps from y -> p(z|y), which gives an inferred {Mean, Std} of the prior distribution of the **specific** cluster from which the image came. To find LL, I need such {Mean, Std} pairs for ALL the clusters. I'm wondering what's the right way to do that?

**Comment by mazrk7** [comment]
No problem!

Yes, the fully-connected layer y -> p(z|y) does only infer the parameters of the prior associated with a specific cluster (i.e. y=k). But then you could also iterate over each y value, as in https://github.com/mazrk7/gmvae/blob/master/scripts/gmvae.py#L152 when no mixture component indices are provided (clusters=None), so as to compute your p(z) log-likelihood. However, the prior distributions are not conditioned on the input images, so I think for your application you might want to iterate over all y values when sampling from the q(z|x,y) distribution and compute the log-likelihood for the q(z|x) mixture.

Does this help?

**Comment by aayushARM** [comment]
Hey, 
Sorry for the late reply! I got stuck in some other unavoidable work. 
And thanks for clarifying that I can iterate for all y over p(z|y) values to find p(z), I'll try this out.
Regarding your suggestion to "iterate over all y values when sampling from the q(z|x,y) distribution and compute the log-likelihood for the q(z|x) mixture", do you mean if y can take 10 different values, I sample 10 z values, and after that compute the LL of each z w.r.t to the corresponding q(z|x,y) and sum them all up? This will be like finding LL of a sample w.r.t the distribution it was sampled from in the first place, which doesn't make sense. Maybe I'm missing something here?

**Comment by mazrk7** [comment]
Hey, no sorry I mean sample z once from q(z|x,y) and then use the _logits_ or probabilities of the q(y|x) distribution to represent all y values. You can then just take the log sum of your log probabilities for the sampled z, scaled by the mixture probabilities, as in your earlier equation.

In case that's not clear, maybe have a look at how the TensorFlow Probability library does it: https://github.com/tensorflow/probability/blob/v0.11.1/tensorflow_probability/python/distributions/mixture_same_family.py#L346

**Comment by mazrk7** [comment]
Hi @aayushARM, is it alright if I close the issue?

**Comment by aayushARM** [comment]
Hey, yeah the issue is largely solved at this point. I'm closing it. Thanks for your help! :)
